### PR TITLE
buffer: fix writes into out-of-memory buffers

### PR DESCRIPTION
**Pull request by pks-t**
This fixes two issues with `git_buf`:

1. An infinite loop when allocating a huge amount of bytes close to `SIZE_MAX`. While obviously a bogus scenario, I still think the fix is worthwhile.
2. We failed to correctly ensure sizes of OOM'd buffers, which resulted in us writing into the `git_buf__oom` buffer.

Found while investigating #5199 once again. This is probably not the cause for the observed crashes, though, as the stack traces point somewhere else.

### CODE REVIEW

**Comment by ethomson** [comment]
Apparently you can't rebuild until the build finishes completely.  I cancelled & requeued manually.

**Comment by pks-t** [comment]
Thanks! I tried to read the macOS log that was failing, but it really is completely unreadable due to spam caused by leak checks. :/

**Comment by pks-t** [comment]
Oops, there was a memory leak in core::buffer::avoid_printing_into_oom_buffer. `git_buf_dispose` isn't going to do anything sensible if the buffer's pointer was swapped out for `git_buf__oom`.

File: tests/core/buffer.c [hunk 1]
```diff
@@ -1204,3 +1204,37 @@ void test_core_buffer__dont_grow_borrowed(void)
 
 	cl_git_fail_with(GIT_EINVALID, git_buf_grow(&buf, 1024));
 }
+
+void test_core_buffer__dont_hit_infinite_loop_when_resizing(void)
+{
+	git_buf buf = GIT_BUF_INIT;
+
+	cl_git_pass(git_buf_puts(&buf, "foobar"));
+	cl_git_fail(git_buf_try_grow(&buf, SIZE_MAX, true));
```
Comment (ethomson) [inline tests/core/buffer.c]: It looks like macOS isn't behaving the way you want.  I think that you're relying on allocation semantics here and that you may need to substitute an intentionally failing allocator here that asserts that it ends up with `SIZE_MAX` and fails appropriately.

Comment (pks-t) [reply tests/core/buffer.c]: Yeah, I expected this to happen but hoped it wouldn't, so I just gave it a try. I've fixed the tests to not rely on allocation semantics

**Comment by pks-t** [comment]
Finally green with another memory leak fix in the OOM case.

**Pull request merged**